src/components/PointCloudViewer.tsx

```diff
@@ -0,0 +1,177 @@
+import React, { useEffect, useMemo, useRef, useState } from 'react';
+import { Canvas } from 'react-three-fiber';
+import { Matrix4, Vector3 } from 'three';
+import { fb } from '../schema';
+
+const getFloatFromDataArray = (
+  arr: Uint8Array,
+  start: number,
+  end: number,
+  little_endian: boolean
+) => {
+  let view = new DataView(arr.slice(start, end).buffer);
+
+  return view.getFloat32(0, little_endian);
+};
+
+export default function PointCloudViewer(props: {
+  data: flatbuffers.ByteBuffer;
+  enabled: boolean;
+  point_skip_ratio: number;
+  point_cloud_scale: number;
+  max_intensity: number;
+  large: boolean;
+}) {
+  const {
+    point_skip_ratio,
+    point_cloud_scale,
+    max_intensity,
+    data,
+    large,
+  } = props;
+  const [pointData, setPointData] = useState(new Float32Array(0));
+  const [pointIntensityData, setPointIntensityData] = useState(
+    new Float32Array(0)
+  );
+
+  useEffect(() => {
+    const cloud = fb.sensor_msgs.PointCloud2.getRootAsPointCloud2(data);
+
+    // Identify the x, y, z fields from the cloud
+    let xOffset: number | undefined;
+    let yOffset: number | undefined;
+    let zOffset: number | undefined;
+    let intensityOffset: number | undefined;
+
+    for (let i = 0; i < cloud.fieldsLength(); i++) {
+      if (cloud.fields(i)?.name() === 'x') {
+        xOffset = cloud.fields(i)?.offset();
+      } else if (cloud.fields(i)?.name() === 'y') {
+        yOffset = cloud.fields(i)?.offset();
+      } else if (cloud.fields(i)?.name() === 'z') {
+        zOffset = cloud.fields(i)?.offset();
+      } else if (cloud.fields(i)?.name() === 'intensity') {
+        intensityOffset = cloud.fields(i)?.offset();
+      }
+    }
+
+    if (
+      xOffset === undefined ||
+      yOffset === undefined ||
+      zOffset === undefined
+    ) {
+      console.warn('Unable to find required fields in point cloud data.');
+      return;
+    }
+
+    const step = cloud.pointStep();
+    let num_points = cloud.dataLength() / cloud.pointStep();
+    num_points = num_points / point_skip_ratio;
+    const points = new Float32Array(num_points * 3);
+    const pointIntensity = new Float32Array(num_points * 3);
+
+    for (let i = 0; i < num_points; i += point_skip_ratio) {
+      const cloud_idx = i * step;
+      const point_idx = i * 3;
+
+      // assume fields are of length 4, and are float datatype
+      points[point_idx + 0] = getFloatFromDataArray(
+        cloud.dataArray()!,
+        cloud_idx + xOffset,
+        cloud_idx + xOffset + 4,
+        !cloud.isBigendian()
+      );
```

Comment:
You might be able to simplify this a little:
```suggestion
    const dataView = new DataView(cloud.dataArray()!);
    for (let i = 0; i < num_points; i += point_skip_ratio) {
      const cloud_idx = i * step;
      const point_idx = i * 3;

      // assume fields are of length 4, and are float datatype
      points[point_idx + 0] = dataView.getFloat32(
        cloud_idx + xOffset,
        !cloud.isBigendian()
      );
```
My main concern is just that the `slice()` call in `getFloatFromDataArray` allocates an object, and this seems like a tight loop.

Reply:
Sounds good, I'll test this out and see how it works

Reply:
Turns out I had to do a little more magic here to figure out the right byte offset because of JS being annoying with types, but the new method (using `subarray`), doesn't copy any data, just creates a new pointer to the right place in the array.